File: 21. Merge Two Sorted Lists.py
# use dummyHead and a pointer

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode(0)
        curr = dummyHead
        
        while l1 and l2:
            v1 = l1.val
            v2 = l2.val
            
            if v1 < v2:
                curr.next = ListNode(v1)
                l1 = l1.next
            else:
                curr.next = ListNode(v2)
                l2 = l2.next
            curr = curr.next
        
        if l1: curr.next = l1
        if l2: curr.next = l2
        
        return dummyHead.next

File: 21. Merge Two Sorted Lists.cpp
// https://leetcode.com/problems/merge-two-sorted-lists/

// Merge two sorted linked lists and return it as a sorted list. The list should be
// made by splicing together the nodes of the first two lists.

////////////////////////////////////////////////////////////////////////////////

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *dummy = new ListNode(), *curr = dummy;
        while (l1 != nullptr and l2 != nullptr) {
            if (l1->val < l2->val) {
                curr->next = new ListNode(l1->val);
                curr = curr->next; l1 = l1->next;
            } else {
                curr->next = new ListNode(l2->val);
                curr = curr->next; l2 = l2->next;
            }
        }
        if (l1 != nullptr) curr->next = l1;
        if (l2 != nullptr) curr->next = l2;
        return dummy->next;
    }
};
